### PR TITLE
[11.0][NEW] website_sale_product_detail_attribute_image: new module

### PR DESCRIPTION
**Pull request by sergio-teruel**
New module for display attribute image in produts detail shop view
cc @Tecnativa

### CODE REVIEW

**Comment by sergio-teruel** [comment]
I have added free images for demo data from https://www.iconfinder.com/, Where I put this images information?

**Comment by pedrobaeza** [comment]
Put them on CREDITS.rst

File: website_sale_product_detail_attribute_image/readme/DESCRIPTION.rst [hunk 1]
```diff
@@ -0,0 +1,2 @@
+This module extends the functionality of website sale module to allow display
```
Comment (pedrobaeza) [inline website_sale_product_detail_attribute_image/readme/DESCRIPTION.rst]: s/to allow/to allow to

File: website_sale_product_detail_attribute_image/readme/DESCRIPTION.rst [hunk 2]
```diff
@@ -0,0 +1,2 @@
...
+product attributes images in shop product detail.
```
Comment (pedrobaeza) [inline website_sale_product_detail_attribute_image/readme/DESCRIPTION.rst]: s/product attributes images in shop product detail/images related to product attributes in e-commerce product page

File: website_sale_product_detail_attribute_image/models/product_attribute.py [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+
+
+class ProductAttribute(models.Model):
+    _inherit = 'product.attribute'
+
+    website_product_detail_image = fields.Binary(
+        string='Website detail image',
+        attachment=True,
+        help='Small-sized image of the attribute for shop online product '
+             'detail. It is automatically resized as a 64x64 px image, with '
```
Comment (pedrobaeza) [inline website_sale_product_detail_attribute_image/models/product_attribute.py]: You shouldn't resize this automatically. What about if someone wants it to 128x128? If automatic resize, you have to put a configuration option to select the preferred size.

File: website_sale_product_detail_attribute_image/models/product_attribute.py [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+             'aspect ratio preserved.',
+    )
+    website_product_detail_display = fields.Boolean(
```
Comment (pedrobaeza) [inline website_sale_product_detail_attribute_image/models/product_attribute.py]: Being homogeneous with other namings, I would call this `website_product_detail_image_published`

File: website_sale_product_detail_attribute_image/controllers/main.py [hunk 5]
```diff
@@ -0,0 +1,16 @@
...
+
+
+class WebsiteSaleProductDetailAttributeImage(WebsiteSale):
+
+    @http.route()
+    def product(self, product, category='', search='', **kwargs):
+        res = super().product(
+            product, category=category, search=search, **kwargs)
+        image_attributes = product.attribute_line_ids.filtered(
+            lambda x: x.attribute_id.website_product_detail_display)
```
Comment (pedrobaeza) [inline website_sale_product_detail_attribute_image/controllers/main.py]: Isn't better to populate images directly:
```python
attribute_images = product.attribute_line_ids.filtered(
    lambda x: x.attribute_id.website_product_detail_display
).mapped('website_product_detail_image')
```

File: website_sale_product_detail_attribute_image/static/src/less/style.less [hunk 6]
```diff
@@ -0,0 +1,8 @@
+.product-detail-attributes img {
+    width: 40px;
+    height: 40px;
+}
+.product-detail-attributes div {
+    font-size: 85%;
+    /*line-height: 1;*/
```
Comment (pedrobaeza) [inline website_sale_product_detail_attribute_image/static/src/less/style.less]: Remove this if not used

File: website_sale_product_detail_attribute_image/tests/test_product_detail_attribute_image.py [hunk 7]
```diff
@@ -0,0 +1,28 @@
+# Copyright 2019 Tecnativa - Sergio Teruel
+# License AGPL-3.0 or later (https://www.gnu.org/licenses/agpl).
+from odoo.tests.common import SavepointCase
+
+
+class TestProductAttribute(SavepointCase):
+    at_install = False
+    post_install = True
```
Comment (pedrobaeza) [inline website_sale_product_detail_attribute_image/tests/test_product_detail_attribute_image.py]: Is this needed?

File: website_sale_product_detail_attribute_image/tests/test_product_detail_attribute_image.py [hunk 8]
```diff
@@ -0,0 +1,28 @@
...
+        attribute = self.env['product.attribute'].create({
+            'name': 'Pictogram attribute2',
+            'website_product_detail_image': self.image,
+        })
+        self.assertEqual(attribute.website_product_detail_image, self.image)
+
+    def test_write_image_for_attribute(self):
+        self.attribute.website_product_detail_image = self.image
+        self.assertEqual(
+            self.attribute.website_product_detail_image, self.image)
```
Comment (pedrobaeza) [inline website_sale_product_detail_attribute_image/tests/test_product_detail_attribute_image.py]: This is not testing actually the resize

File: website_sale_product_detail_attribute_image/views/product_attribute_views.xml [hunk 9]
```diff
@@ -0,0 +1,30 @@
...
+<!-- Copyright 2019 Tecnativa - Sergio Teruel
+     License AGPL-3.0 or later (https://www.gnu.org/licenses/agpl). -->
+<odoo>
+
+    <record id="attribute_tree_view" model="ir.ui.view">
+        <field name="model">product.attribute</field>
+        <field name="inherit_id" ref="product.attribute_tree_view"/>
+        <field name="arch" type="xml">
+            <field name="create_variant" position="after">
+                <field name="website_product_detail_display"/>
```
Comment (pedrobaeza) [inline website_sale_product_detail_attribute_image/views/product_attribute_views.xml]: widget switch or boolean?

File: website_sale_product_detail_attribute_image/controllers/main.py (hunk 5 above)
Comment (pedrobaeza) [reply website_sale_product_detail_attribute_image/controllers/main.py]: I self-answer: you are also displaying the attribute name. Not sure about this being coupled between backend and frontend. It's better to have an extra translate field for website display, but fallbacking to this one if not set.

File: website_sale_product_detail_attribute_image/views/templates.xml [hunk 10]
```diff
@@ -0,0 +1,25 @@
...
+            <div class="row mt16 mb0 product-detail-attributes">
+                <t t-foreach="image_attributes" t-as="image_attribute">
+                    <div class="col-md-2">
+                        <div class="text-center"> <span t-field="image_attribute.attribute_id.name"/> </div>
+                        <div class="text-center">
+                            <img class="img"
+                                 t-attf-src="/web/image/product.attribute/#{image_attribute.attribute_id.id}/website_product_detail_image"/>
+                        </div>
+                        <t t-foreach="image_attribute.value_ids" t-as="value">
+                            <div class="text-center"> <span t-field="value.name"/> </div>
```
Comment (pedrobaeza) [inline website_sale_product_detail_attribute_image/views/templates.xml]: The same about showing these backend string here. You are not mentioning this in the README, nor the label of the attribute.

**Review by pedrobaeza — CHANGES_REQUESTED** [review]

**Comment by sergio-teruel** [comment]
Changes done.

File: website_sale_product_detail_attribute_image/views/product_attribute_views.xml [hunk 11]
```diff
@@ -0,0 +1,52 @@
...
+<!-- Copyright 2019 Tecnativa - Sergio Teruel
+     License AGPL-3.0 or later (https://www.gnu.org/licenses/agpl). -->
+<odoo>
+
+    <record id="attribute_tree_view" model="ir.ui.view">
+        <field name="model">product.attribute</field>
+        <field name="inherit_id" ref="product.attribute_tree_view"/>
+        <field name="arch" type="xml">
+            <field name="create_variant" position="after">
+                <field name="website_product_detail_image"/>
```
Comment (Tardo) [inline website_sale_product_detail_attribute_image/views/product_attribute_views.xml:11]: Why not uses widget='image' here? size?

File: website_sale_product_detail_attribute_image/views/product_attribute_views.xml [hunk 12]
```diff
@@ -0,0 +1,52 @@
...
+
+    <record id="product_attribute_view_form" model="ir.ui.view">
+        <field name="model">product.attribute</field>
+        <field name="inherit_id" ref="product.product_attribute_view_form"/>
+        <field name="arch" type="xml">
+            <group name="main_fields" position="after">
+                <group name="product_detail_image" string="Website product image detail">
+                    <field name="website_product_detail_image" widget="image"/>
+                    <field name="website_name"/>
+                    <field name="website_product_detail_image_published"/>
```
Comment (Tardo) [inline website_sale_product_detail_attribute_image/views/product_attribute_views.xml]: Why not uses widget="boolean_toggle" here?

File: website_sale_product_detail_attribute_image/views/product_attribute_views.xml (hunk 11 above)
Comment (pedrobaeza) [reply website_sale_product_detail_attribute_image/views/product_attribute_views.xml:11]: is it working on trees?

Comment (Tardo) [reply website_sale_product_detail_attribute_image/views/product_attribute_views.xml:11]: Work but without resize image :/ perhaps can use: https://github.com/OCA/web/tree/11.0/web_tree_image

Comment (pedrobaeza) [reply website_sale_product_detail_attribute_image/views/product_attribute_views.xml:11]: Well, maybe too much overhead for this, don't you think? Let Sergio decides.

Comment (sergio-teruel) [reply website_sale_product_detail_attribute_image/views/product_attribute_views.xml:11]: I don't want dependencies of other modules, If the user wants to view the image only has to do click on record, also this model, once it has been configured the user not will coming back to see the images every day.
But I thought so too :smile:

File: website_sale_product_detail_attribute_image/views/product_attribute_views.xml (hunk 12 above)
Comment (sergio-teruel) [reply website_sale_product_detail_attribute_image/views/product_attribute_views.xml]: Yes!!.. It slipped my mind!!

**Comment by sergio-teruel** [comment]
Changes done

**Review by pedrobaeza — APPROVED** [review]

**Review by Tardo — APPROVED** [review]
LGTM, tested on runbot

**Review by rafaelbn — APPROVED** [review]
Tested in runbot :+1:

**Pull request merged**